R/busy-indicators.R

```diff
@@ -0,0 +1,184 @@
+  tags$script(js)
+}
+
+#' Customize busy indicator options.
+#'
+#' To customize the appearance of the busy indicators, include the result of
+#' this function in the app's UI.
+#'
+#'
+#' @param ... Currently ignored (for future expansion).
+#' @param spinner_color The color of the spinner. This can be any valid CSS
+#'   color. Defaults to the app's "primary" color (if Bootstrap is on the page).
+#' @param spinner_size The size of the spinner. This can be any valid CSS size.
+#' @param spinner_delay The amount of time to wait before showing the spinner.
+#'   This can be any valid CSS time and can useful for not showing the spinner
+#'   if the computation finishes quickly.
+#' @param pulse_background A CCS background definition for the pulse. The
+#'   default uses a
+#'   [linear-gradient](https://developer.mozilla.org/en-US/docs/Web/CSS/gradient/linear-gradient)
+#'   of the theme's indigo, purple, and pink colors.
+#' @param pulse_height The height of the pulsing banner. This can be any valid
+#'   CSS size.
+#' @param pulse_speed The speed of the pulsing banner. This can be any valid CSS
+#'   time.
+#'
+#' @export
+#' @seealso [useBusyIndicators()] for enabling/disabling busy indicators.
+#' @examplesIf interactive()
+#'
+#' library(bslib)
+#'
+#' ui <- page_fillable(
+#'   useBusyIndicators(),
+#'   busyIndicatorOptions(spinner_color = "orange"),
+#'   card(
+#'     card_header(
+#'       "A plot",
+#'       input_task_button("simulate", "Simulate"),
+#'       class = "d-flex justify-content-between align-items-center"
+#'     ),
+#'     plotOutput("p"),
+#'   )
+#' )
+#'
+#' server <- function(input, output) {
+#'   output$p <- renderPlot({
+#'     input$simulate
+#'     Sys.sleep(4)
+#'     plot(x = rnorm(100), y = rnorm(100))
+#'   })
+#' }
+#'
+#' shinyApp(ui, server)
+busyIndicatorOptions <- function(
+  ...,
+  spinner_color = NULL,
+  spinner_size = NULL,
+  spinner_delay = NULL,
+  pulse_background = NULL,
+  pulse_height = NULL,
+  pulse_speed = NULL
+) {
+
+  rlang::check_dots_empty()
+
+  res <- tagList(
+    spinnerOptions(
+      color = spinner_color,
+      size = spinner_size,
+      delay = spinner_delay
+    ),
+    pulseOptions(
+      background = pulse_background,
+      height = pulse_height,
+      speed = pulse_speed
+    )
+  )
+
+  dropNulls(res)
+}
+
+
+# TODO: allow for customization of the spinner type.
+spinnerOptions <- function(color = NULL, size = NULL, delay = NULL) {
+  if (is.null(color) && is.null(size) && is.null(delay))  {
+    return(NULL)
+  }
+
+  # Options controlled via CSS variables.
+  css_vars <- paste0(c(
+    if (!is.null(color)) sprintf("--shiny-spinner-color: %s", htmltools::parseCssColors(color)),
+    if (!is.null(size)) sprintf("--shiny-spinner-size: %s", htmltools::validateCssUnit(size)),
+    if (!is.null(delay)) sprintf("--shiny-spinner-delay: %s", delay)
+  ), collapse = ";")
+
+  tags$style(HTML(paste0(":root {", css_vars, "}")))
```

Comment:
I originally had that, but I ended removing it, since it's only relevant for spinners, and thinking about customization as just a just simple "global" settings, means it makes more sense to have a singular `busyIndicatorOptions()`, as opposed to spinner/pulse specific options. 

How about I just add a section to the docs about "scoping" customizations via CSS variables?

Reply:
1. There are cases where the default will be the wrong choice, e.g. on cards with `class="text-bg-primary"`
2. It's easier to add `selector` and document it (and document that it's only relevant for spinners) than it is to add a section describing all of our CSS variables and how to scope them.